jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -215,5 +697,1697 @@ public void clear()
+        }
+
+        @Override
+        public Mutable add(ByteBuffer bytes) throws ReadOnlyBufferException, BufferOverflowException
+        {
+            getWrapped().asMutable().add(bytes);
+            return this;
+        }
+
+        @Override
+        public Mutable add(RetainableByteBuffer bytes) throws ReadOnlyBufferException, BufferOverflowException
+        {
+            getWrapped().asMutable().add(bytes);
+            return this;
+        }
+
+        @Override
+        public Mutable put(byte b)
+        {
+            getWrapped().asMutable().put(b);
+            return this;
+        }
+
+        @Override
+        public Mutable put(long index, byte b)
+        {
+            getWrapped().asMutable().put(index, b);
+            return this;
+        }
+
+        @Override
+        public Mutable putShort(short s)
+        {
+            getWrapped().asMutable().putShort(s);
+            return this;
+        }
+
+        @Override
+        public Mutable putInt(int i)
+        {
+            getWrapped().asMutable().putInt(i);
+            return this;
+        }
+
+        @Override
+        public Mutable putLong(long l)
+        {
+            getWrapped().asMutable().putLong(l);
+            return this;
+        }
+
+        @Override
+        public Mutable put(byte[] bytes, int offset, int length)
+        {
+            getWrapped().asMutable().put(bytes, offset, length);
+            return this;
+        }
+
+        @Override
+        public String toDetailString()
+        {
+            return getWrapped().toDetailString();
+        }
+    }
+
+    /**
+     * An abstract implementation of {@link RetainableByteBuffer} that provides the basic {@link Retainable} functionality
+     */
+    abstract class Abstract extends Retainable.Wrapper implements Mutable
+    {
+        public Abstract()
+        {
+            this(new ReferenceCounter());
+        }
+
+        public Abstract(Retainable retainable)
+        {
+            super(retainable);
+        }
+
+        /**
+         * @return A string showing the info about this buffer
+         */
+        @Override
+        public String toString()
+        {
+            StringBuilder builder = new StringBuilder();
+            addStringInfo(builder);
+            return builder.toString();
+        }
+
+        /**
+         * @return A string showing the info and detail about this buffer
+         */
+        @Override
+        public String toDetailString()
+        {
+            StringBuilder builder = new StringBuilder();
+            addStringInfo(builder);
+            builder.append("={");
+            addValueString(builder);
+            builder.append("}");
+            return builder.toString();
+        }
+
+        protected void addStringInfo(StringBuilder builder)
+        {
+            builder.append(getClass().getSimpleName());
+            builder.append("@");
+            builder.append(Integer.toHexString(System.identityHashCode(this)));
```

Comment:
Bttetr use `hashcode()` instead of `System.identityHashCode(this)`.

Reply:
I'm a bit worried that somebody might write a buffer impl that hashes the contents of the buffer.   We don't do that, but this feels safer.